File: Level3/11.가장먼노드/3.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <string>
#include <vector>
#define MAX 20001

using namespace std;

const int INF = 2e9;

struct Node {
  int idx, dist;
  bool operator<(const Node &b) const { return dist > b.dist; }
};

int solution(int n, vector<vector<int>> edge) {
  int answer = 0, maxDist = 0;

  priority_queue<Node> pq;
  int cache[MAX];
  for (int i = 1; i <= n; i++) cache[i] = INF;
  vector<int> graph[MAX];

  for (const auto &el : edge) {
    int a = el[0], b = el[1];
    graph[a].push_back(b);
    graph[b].push_back(a);
  }

  cache[1] = 0;
  pq.push({1, 0});
  while (!pq.empty()) {
    int idx = pq.top().idx;
    int dist = pq.top().dist;
    pq.pop();

    if (dist > cache[idx]) continue;

    for (const auto &next : graph[idx]) {
      if (dist + 1 >= cache[next]) continue;

      cache[next] = dist + 1;
      pq.push({next, dist + 1});

      if (cache[next] > maxDist) maxDist = cache[next];
    }
  }
  for (int i = 2; i <= n; i++) {
    if (cache[i] == maxDist) answer++;
  }

  return answer;
}

File: Level3/합승택시요금/3.js
function solution(n, s, a, b, fares) {
  const INF = 987654321;
  const graph = [[]];

  for (let i = 1; i <= n; i++) {
    graph[i] = [0];
    for (let j = 1; j <= n; j++) {
      if (i === j) graph[i][j] = 0;
      else graph[i][j] = INF;
    }
  }

  fares.forEach(([c, d, f]) => {
    graph[c][d] = f;
    graph[d][c] = f;
  });

  for (let k = 1; k <= n; k++) {
    for (let i = 1; i <= n; i++) {
      for (let j = 1; j <= n; j++) {
        graph[i][j] = Math.min(graph[i][j], graph[i][k] + graph[k][j]);
      }
    }
  }

  let answer = graph[s][a] + graph[s][b];

  for (let i = 1; i <= n; i++) {
    answer = Math.min(answer, graph[s][i] + graph[i][a] + graph[i][b]);
  }

  return answer;
}
